### PR TITLE
QL: Refactor function resolution strategy

### PR DESCRIPTION
**Pull request by costin**
The SQL strategy for resolving functions is still used inside QL.
This commit simplifies it by breaking the enum into a pluggable
interface that SQL extends accordingly.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java [hunk 1]
```diff
@@ -118,18 +118,21 @@ public boolean functionExists(String functionName) {
...
     }
 
     /**
      * Build a {@linkplain FunctionDefinition} for a no-argument function that
      * is not aware of time zone and does not support {@code DISTINCT}.
      */
-    protected static <T extends Function> FunctionDefinition def(Class<T> function,
-            java.util.function.Function<Source, T> ctorRef, String... names) {
+    public static <T extends Function> FunctionDefinition def(Class<T> function,
+                                                              java.util.function.Function<Source, T> ctorRef, String... names) {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:135]: This class needs to be refactored in a follow-up. The methods were made public from protected to make them a bit consistent - some methods were public but the interface in the arguments was protected while methods without interfaces were protected...

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java [hunk 2]
```diff
@@ -247,7 +250,7 @@ public boolean functionExists(String functionName) {
         return def(function, builder, false, names);
     }
 
-    protected interface DistinctAwareUnaryFunctionBuilder<T> {
+    public interface DistinctAwareUnaryFunctionBuilder<T> {
```
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:253]: protected interface as an argument to a public method.

**Comment by costin** [comment]
For context, this refactoring was started by looking at introducing case insensitive functions in EQL.

**Comment by costin** [comment]
@elasticmachine update branch

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+        // No children aren't supported
+        e = expectThrows(ParsingException.class, () ->
+            uf(DEFAULT).buildResolved(randomConfiguration(), def));
+        assertThat(e.getMessage(), endsWith("expects exactly one argument"));
+
+        // Multiple children aren't supported
+        e = expectThrows(ParsingException.class, () ->
+            uf(DEFAULT, mock(Expression.class), mock(Expression.class)).buildResolved(randomConfiguration(), def));
+        assertThat(e.getMessage(), endsWith("expects exactly one argument"));
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:73]: These tests are duplicates of the method with the same name in `FunctionRegistryTests` in QL.   Why duplicating these?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    public void testUnaryFunction() {
+        UnresolvedFunction ur = uf(DEFAULT, mock(Expression.class));
+        FunctionRegistry r = new FunctionRegistry(def(DummyFunction.class, (Source l, Expression e) -> {
+            assertSame(e, ur.children().get(0));
+            return new DummyFunction(l);
+        }, "DUMMY_FUNCTION"));
+        FunctionDefinition def = r.resolveFunction(ur.name());
+        assertFalse(def.extractViable());
+        assertEquals(ur.source(), ur.buildResolved(randomConfiguration(), def).source());
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:58]: This test is a duplicate of a test inside `FunctionRegistryTests` in QL.

**Review by astefan — COMMENTED** [review]
Left a comment regarding some tests.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:73]: The methods in the base class don't contain all the strategies - the first lines are code are duplicated (the creation is needed, plus the two assertions) but the rest in the subclass are not.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:58]: `FunctionRegistryTests`:

```
    public void testNoArgFunction() {
        UnresolvedFunction ur = uf(DEFAULT);
        FunctionRegistry r = new FunctionRegistry(def(DummyFunction.class, DummyFunction::new, "DUMMY_FUNCTION"));
        FunctionDefinition def = r.resolveFunction(ur.name());
        assertEquals(ur.source(), ur.buildResolved(randomConfiguration(), def).source());
    }
```

`SqlFunctionRegistryTests`:

```
    public void testNoArgFunction() {
        UnresolvedFunction ur = uf(DEFAULT);
        FunctionRegistry r = new FunctionRegistry(def(DummyFunction.class, DummyFunction::new, "DUMMY_FUNCTION"));
        FunctionDefinition def = r.resolveFunction(ur.name());
        assertEquals(ur.source(), ur.buildResolved(randomConfiguration(), def).source());

        // Distinct isn't supported
        ParsingException e = expectThrows(ParsingException.class, () ->
            uf(DISTINCT).buildResolved(randomConfiguration(), def));
        assertThat(e.getMessage(), endsWith("does not support DISTINCT yet it was specified"));

        // Any children aren't supported
        e = expectThrows(ParsingException.class, () ->
            uf(DEFAULT, mock(Expression.class)).buildResolved(randomConfiguration(), def));
        assertThat(e.getMessage(), endsWith("expects no arguments"));
    }
```

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:73]: You could keep only the initialization part and the DISTINCT test, no?

Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistryTests.java:73]: I'll look further at eliminating the extra assertions.

**Review by matriv — APPROVED** [review]
LGTM, other that the test duplication (assertions actually) that would be nice to avoid if possible.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionDefinition.java [hunk 5]
```diff
@@ -23,18 +23,22 @@
     private final String name;
     private final List<String> aliases;
     private final Class<? extends Function> clazz;
+    private final Builder builder;
+
     /**
      * Is this a datetime function compatible with {@code EXTRACT}.
      */
+    // TODO: needs refactoring
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionDefinition.java:31]: comments always welcome!
I'd only add that adding more details might help anyone else picking this up. (but i favour existence of the comment over its completeness.)

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java [hunk 6]
```diff
@@ -118,18 +118,21 @@ public boolean functionExists(String functionName) {
...
-                .filter(e -> p == null || p.matcher(e.getKey()).matches())
-                .map(e -> new FunctionDefinition(e.getKey(), emptyList(),
-                        e.getValue().clazz(), e.getValue().extractViable(), e.getValue().builder()))
-                .collect(toList());
+            .filter(e -> p == null || p.matcher(e.getKey()).matches())
+            .map(e -> cloneDefinition(e.getKey(), e.getValue()))
+            .collect(toList());
+    }
+
+    protected FunctionDefinition cloneDefinition(String name, FunctionDefinition definition) {
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:126]: I find this approach great, clearer intent in the immutable objects strategy. (_winking towards replaceChildren*_).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionResolutionStrategy.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * Build the real function from this one and resolution metadata.
+     */
+    default Function buildResolved(UnresolvedFunction uf, Configuration cfg, FunctionDefinition def) {
+        return def.builder().build(uf, false, cfg);
+    }
+
+    /**
+     * The name of the kind of thing being resolved. Used when
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionResolutionStrategy.java:31]: 1st sentence might be worth a rephrase. (`What kind of thing is being resolved?`?)

**Review by bpintea — APPROVED** [review]
LGTM.

**Pull request merged**